core/src/main/java/org/elasticsearch/cluster/metadata/IndexGraveyard.java

```diff
@@ -0,0 +1,459 @@
+
+    @Override
+    public void writeTo(final StreamOutput out) throws IOException {
+        out.writeVInt(tombstones.size());
+        for (Tombstone tombstone : tombstones) {
+            tombstone.writeTo(out);
+        }
+    }
+
+    @Override
+    public IndexGraveyard readFrom(final StreamInput in) throws IOException {
+        return new IndexGraveyard(in);
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public Diff<MetaData.Custom> diff(final MetaData.Custom previous) {
+        return new IndexGraveyardDiff((IndexGraveyard) previous, this);
+    }
+
+    @Override
+    public Diff<MetaData.Custom> readDiffFrom(final StreamInput in) throws IOException {
+        return new IndexGraveyardDiff(in);
+    }
+
+    public static IndexGraveyard.Builder builder() {
+        return new IndexGraveyard.Builder();
+    }
+
+    public static IndexGraveyard.Builder builder(final IndexGraveyard graveyard) {
+        return new IndexGraveyard.Builder(graveyard);
+    }
+
+    /**
+     * A class to build an IndexGraveyard.
+     */
+    final public static class Builder {
+        private List<Tombstone> tombstones;
+        private int numPurged = -1;
+
+        private Builder() {
+            tombstones = new ArrayList<>();
+        }
+
+        private Builder(IndexGraveyard that) {
+            tombstones = new ArrayList<>(that.getTombstones());
+        }
+
+        /**
+         * A copy of the current tombstones in the builder.
+         */
+        public List<Tombstone> tombstones() {
+            return Collections.unmodifiableList(tombstones);
+        }
+
+        /**
+         * Add a deleted index to the list of tombstones in the cluster state.
+         */
+        public Builder addTombstone(final Index index) {
+            final Tombstone tombstone = new Tombstone(index, System.currentTimeMillis());
+            tombstones.add(tombstone);
+            return this;
+        }
+
+        /**
+         * Add a list of tombstones to the graveyard.
+         */
+        Builder addTombstones(final List<Tombstone> tombstones) {
+            this.tombstones.addAll(tombstones);
+            return this;
+        }
+
+        /**
+         * Get the number of tombstones that were purged.  This should *only* be called
+         * after build() has been called.
+         */
+        public int getNumPurged() {
+            assert numPurged != -1;
+            return numPurged;
+        }
+
+        /**
+         * Purge tombstone entries.  Returns the number of entries that were purged.
+         *
+         * Tombstones are purged if the number of tombstones in the list
+         * is greater than the input parameter of maximum allowed tombstones.
+         * Tombstones are purged until the list is equal to the maximum allowed.
+         */
+        private int purge(final int maxTombstones) {
+            int numPurged = tombstones().size() - maxTombstones;
+            if (numPurged <= 0) {
+                return 0;
```

Comment:
Oh it's local, in that case, please name it something else, it shadows the class var `numPurged`